docs/adapter_impl_notes.md

```diff
@@ -0,0 +1,184 @@
+# Adapter Implementation Notes
+
+This page talks about implementation details of writing adapters. If you want to write your own adapter for your own backend, use this document as a guide.
+
+For this document, we'll use the Gmail Adapter as an example. In this case, the backend consists a set of email messages. Each message contains a timestamp (when the message was received) and fields (email headers such as ``From:``, ``To:``, ``Subject:``, the message body, etc). The Gmail API supports [search expressions](https://support.google.com/mail/answer/7190?hl=en) that select messages based on date, headers, or a free-text search on the message contents.
+
+As of 2015-12-23, the adapter implementation doesn't *exactly* match this description. For example, the adapter doesn't yet support filtering expressions or live programs. However, the notes still do a good job of illustrating the concepts.
+
+The adapter's job is to interpret the options included in the juttle ``read`` command into a set of matching email messages, construct json objects representing those messages, and pass them to the Juttle Runtime by calling ``emit()``.
+
+More sophisticated adapters work together with the juttle optimizer to push aggregation operations directly into the backend. For example, to count the number of messages in a given time period you could simply fetch all the messages and have the juttle program perform the counting. But it would be more efficient to count the number of messages via Gmail APIs and simply return the count instead.
+
+## Live vs Historical
+
+Juttle's ``read`` command specifies a timerange based on the values of the ``-from`` and ``-to`` options. When the ``-to`` option is less than the current time, an adapter need only fetch the matching set of messages for the provided timerange. Such a program is called *historical*. When the ``-to`` option is greater than the current time, the adapter must also watch for new messages and pass them to the program. Such a problem is called *live*. A program can be both live and historical, with a ``-from`` in the past and a ``-to`` in the future. In that case, the adapter must fetch both old and newly-arriving messages.
```

Comment:
I've only heard of it anecdotally and don't see it in the docs. It sounds like a crummy term though. Can you show examples of where else it is used?